File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:

    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        out = []

        nums.sort()

        def dfs(i, currentList, total):

            if total == target:
                out.append(list(currentList))
                return
            
            if i >= len(nums):
                return

            if total > target:
                return

            for j in range(i, len(nums)):
                
                if total + nums[j] > target:
                    break

                currentList.append(nums[j])
                dfs(j, currentList, total + nums[j])
                currentList.pop()


        dfs(0, [], 0)

        return out

File: Data Structures & Algorithms/maximum-subarray/submission-2.cpp
class Solution {

    int dfs(int i, bool inSubarray, vector<int>& nums) {
        int n = nums.size();

        if (i == n - 1) {
            if (inSubarray) {
                return max(0, nums[i]);
            } else {
                return nums[i];
            }
        }

        if (inSubarray) {
            return max(
                0,
                nums[i] + dfs(i + 1, true, nums)
            );
        }

        return max(
            dfs(i + 1, false, nums),
            nums[i] + dfs(i + 1, true, nums)
        );
    }

public:
    int maxSubArray(vector<int>& nums) {
        return dfs(0, false, nums);
    }
};
